File: 20200826_Price.js
var mincostTickets = function(days, costs) {
    let lentravel = days[-1];
    let ticket = new Array(lentravel);
    ticket.fill(0);
    for (let i=0; i<days.length;i++) {
        ticket[days[i]-1] = 1;
    }

    function price (n) {
        if (n==-1) return 0
        if (ticket[n] ==0) return price(n-1)
        let a1 = Math.max(-1,n-1);
        let a2 = Math.max(-1,n-7);
        let a3 = Math.max(-1, n-15);
        let kq;
        kq = Math.min(price(a1) + costs[0], price(a2) + costs[1], price(a3) + costs[2]);
        return kq;
    }
    return price(lentravel-1);
};

let days = [1,2,3,4,5,6,7,8,9,10,30,31]; let costs = [2,7,15]

console.log(mincostTickets(days,costs))

File: 20200826 Price.py

def mincostTickets(days, costs):
    NumberOfDay = days[-1]
    tickets = [0 for i in range (NumberOfDay)]
    for ele in days:
        tickets[ele-1]=1
    arr=[0]
    for i in range(NumberOfDay):
        if tickets[i]==0:
            arr.append(arr[i])
        else:
            a0 = max(0,i)
            a1 = max(0,i-6)
            a2 = max(0,i-29)
            arr.append(min(arr[a0]+costs[0],arr[a1]+costs[1],arr[a2]+costs[2]))
    return arr[-1]

days = [1,4,6,7,8,20]
costs = [2,7,15]
print(mincostTickets(days,costs))